reasoner/planner/AnswerCountEstimator.java

```diff
@@ -0,0 +1,486 @@
+                                        long inferredRelationEstimate) {
+                    super(variables);
+                    this.relationTypeEstimate = relationTypeEstimate;
+                    this.rolePlayerEstimates = rolePlayerEstimates;
+                    this.rolePlayerCounts = rolePlayerCounts;
+                    this.inferredRelationEstimate = inferredRelationEstimate;
+                }
+
+                @Override
+                long answerEstimate(Set<Variable> variableFilter) {
+                    long singleRelationEstimate = 1L;
+                    for (TypeVariable key : rolePlayerCounts.keySet()) {
+                        assert rolePlayerEstimates.containsKey(key);
+                        long avgRolePlayers = Double.valueOf(Math.ceil(rolePlayerEstimates.get(key) / relationTypeEstimate)).longValue();
+                        singleRelationEstimate *= nPermuteKforSmallK(avgRolePlayers, rolePlayerCounts.get(key));
+                    }
+
+                    long fullEstimate = Double.valueOf(Math.ceil(relationTypeEstimate * singleRelationEstimate)).longValue() + inferredRelationEstimate;
+                    return fullEstimate;
+                }
+
+                private long nPermuteKforSmallK(long n, long k) {
+                    long ans = 1;
+                    for (int i = 0; i < k; i++) ans *= n - i;
+                    return ans;
+                }
+            }
+        }
+    }
+
+    private static class AnswerCountModel {
+        private final AnswerCountEstimator answerCountEstimator;
+
+        public AnswerCountModel(AnswerCountEstimator answerCountEstimator) {
+            this.answerCountEstimator = answerCountEstimator;
+        }
+
+        private long estimateInferredAnswerCount(Concludable concludable, Set<Variable> variableFilter) {
+            Map<Rule, Set<Unifier>> unifiers = answerCountEstimator.logicMgr.applicableRules(concludable);
+            long inferredEstimate = 0;
+            for (Rule rule : unifiers.keySet()) {
+                for (Unifier unifier : unifiers.get(rule)) {
+                    Set<Identifier.Variable> ruleSideIds = iterate(variableFilter)
+                            .flatMap(v -> iterate(unifier.mapping().get(v.id()))).toSet();
+                    Set<Variable> ruleSideVariables = iterate(ruleSideIds).map(id -> rule.conclusion().conjunction().pattern().variable(id))
+                            .toSet();
+                    if (rule.conclusion().generating().isPresent() && ruleSideVariables.contains(rule.conclusion().generating().get())) {
+                        // There is one generated variable per combination of ALL variables in the conclusion
+                        ruleSideVariables = rule.conclusion().pattern().variables().stream().filter(Variable::isThing).collect(Collectors.toSet());
+                    }
```

Comment:
Generated variables are only in the conclusion and not in the condition - handle that.